### PR TITLE
[OSDEV-1410] Implement POST /api/moderation-events/{moderation_id}/production-locations endpoint

### PR DESCRIPTION
**Pull request by mazursasha1990**
[OSDEV-1410](https://opensupplyhub.atlassian.net/browse/OSDEV-1410) - Moderation. Implement POST /api/moderation-events/{moderation_id}/production-locations endpoint.

In this PR was introduced a new `POST /api/moderation-events/{moderation_id}/production-locations` endpoint.

[OSDEV-1410]: https://opensupplyhub.atlassian.net/browse/OSDEV-1410?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/django/api/views/v1/moderation_events.py [hunk 1]
```diff
@@ -120,3 +126,61 @@ def patch(self, request, pk=None):
...
+                message="Moderation event not found.",
+                status_code=status.HTTP_404_NOT_FOUND,
+            )
+
+        if event.status != ModerationEvent.Status.PENDING:
+            return handle_path_error(
+                field="status",
+                message="Moderation status must be PENDING.",
+                status_code=status.HTTP_410_GONE,
+            )
```
[the remark this answers is not in the document]
Comment (mazursasha1990) [reply src/django/api/views/v1/moderation_events.py:155]: HTTP 410 GONE is suitable for the requirements.

File: src/django/api/moderation_event_actions/approval/add_production_location_strategy.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+    @staticmethod
+    def __create_new_facility(
+        item: FacilityListItem, facility_id: str
+    ) -> Facility:
+        return Facility.objects.create(
+            id=facility_id,
+            name=item.name,
+            address=item.address,
+            country_code=item.country_code,
+            location=item.geocoded_point,
```
[the remark this answers is not in the document]
Comment (mazursasha1990) [reply src/django/api/moderation_event_actions/approval/add_production_location_strategy.py:194]: In this situation, geocoded_point couldn't be None.

File: src/django/api/models/facility/facility_list_item_temp.py [hunk 3]
```diff
@@ -201,7 +201,7 @@ def copy(item):
         FacilityListItemTemp.objects.create(
             id=item.id,
             source=item.source,
-            facility_id=item.facility,
+            facility_id=item.facility_id,
```
Comment (VadimKovalenkoSNF) [inline src/django/api/models/facility/facility_list_item_temp.py:204]: @mazursasha1990 Can you explain pls, why did you do this change? Will it affect functionality implemented in https://opensupplyhub.atlassian.net/browse/OSDEV-507 ?

**Review by VadimKovalenkoSNF — DISMISSED** [review]
Good work, approved. @Innavin369 , pls note that this PR contains some changes related to [OSDEV-1453](https://opensupplyhub.atlassian.net/browse/OSDEV-1453), use them as a reference.

File: src/django/api/views/v1/moderation_events.py [hunk 4]
```diff
@@ -81,26 +87,11 @@ def retrieve(self, _,  pk=None):
 
     @handle_errors_decorator
     def patch(self, request, pk=None):
-        if not (request.user.is_superuser or request.user.is_staff):
-            raise PermissionDenied(
-                detail="Only the Moderator can perform this action."
-            )
+        self.__validate_user_permissions(request)
```
Comment (roman-stolar) [inline src/django/api/views/v1/moderation_events.py:90]: @mazursasha1990 Did you test it after changes?
Also probably we need to discuss and/or update API spec as message will used build-in ViewSet error.

File: src/django/api/views/v1/moderation_events.py [hunk 5]
```diff
@@ -81,26 +87,11 @@ def retrieve(self, _,  pk=None):
...
 
-        if not ModerationIdValidator.is_valid_uuid(pk):
-            return handle_path_error(
-                field="moderation_id",
-                message="Invalid UUID format.",
-                status_code=status.HTTP_400_BAD_REQUEST
-            )
+        self.__validate_uuid(pk)
```
Comment (roman-stolar) [inline src/django/api/views/v1/moderation_events.py:92]: @mazursasha1990 Did you test it after changes?
Also probably we need to discuss and/or update API spec as message will used build-in ViewSet error.

File: src/django/api/views/v1/moderation_events.py [hunk 6]
```diff
@@ -114,9 +105,85 @@ def patch(self, request, pk=None):
...
+        try:
+            item = add_production_location.run_processing()
+        except Exception as error:
+            return self.__handle_processing_error(error)
+
+        return Response(
+            {"os_id": item.facility_id}, status=status.HTTP_201_CREATED
+        )
+
+    @staticmethod
```
Comment (roman-stolar) [inline src/django/api/views/v1/moderation_events.py:138]: I'm just not sure if viewSet is a best way to store private or static methods for endpoint.

Comment (roman-stolar) [reply src/django/api/views/v1/moderation_events.py:138]: Maybe replace it to utils.py or create base viewset for moderation event that covered it.

Comment (roman-stolar) [reply src/django/api/views/v1/moderation_events.py:138]: It will be great, thank you!

Comment (mazursasha1990) [reply src/django/api/views/v1/moderation_events.py]: @roman-stolar I have moved private and static methods to the newly created `ModerationEventsService` class.

Comment (roman-stolar) [reply src/django/api/views/v1/moderation_events.py]: Thank you!

**Review by roman-stolar — APPROVED** [review]
LGTM now.

**Review by roman-stolar — DISMISSED** [review]
Approved

File: src/django/api/views/v1/moderation_events.py [hunk 7]
```diff
@@ -1,28 +1,33 @@
...
+from api.views.v1.opensearch_query_builder.moderation_events_query_builder \
+    import ModerationEventsQueryBuilder
+from api.views.v1.opensearch_query_builder.opensearch_query_director import \
+    OpenSearchQueryDirector
+from api.views.v1.utils import (
+    handle_errors_decorator,
+    serialize_params
+)
+
+log = logging.getLogger(__name__)
```
Comment (VadimKovalenkoSNF) [inline src/django/api/views/v1/moderation_events.py:30]: It looks like this log definition is not using in this `.py` file.

**Review by VadimKovalenkoSNF — CHANGES_REQUESTED** [review]
Minor fix requested.

File: src/django/api/views/v1/moderation_events.py (hunk 7 above)
Comment (mazursasha1990) [reply src/django/api/views/v1/moderation_events.py]: Removed it.
Thank you @VadimKovalenkoSNF.

**Review by VadimKovalenkoSNF — APPROVED** [review]
Approved.

**Review by roman-stolar — APPROVED** [review]

**Review by VadimKovalenkoSNF — APPROVED** [review]
Approve.

**Pull request merged**